manual/parts/threads.tex

```diff
@@ -253,7 +253,10 @@ \subsection{Scheduling Context Donation}
 a notification object with a passive thread blocked waiting on it, the passive thread will receive
 the scheduling context that is bound to the notification object.  The scheduling context is returned
 when the thread blocks on the notification object.  This feature allows for passive servers to use
-notification binding (See \autoref{sec:notification-binding}).
+notification binding (See \autoref{sec:notification-binding}).  If a scheduling context is bound to
+both a notification object and a thread, the behaviour will be the same as for a passive server:
+The scheduling context will be unbound from the thread when it blocks on the bound notification object.
+This is useful when launching passive servers or handling timeout exceptions.
```

Comment:
Could you another sentence here saying why this is useful? Or is this just my lack of detailed understanding here that does not make this obvious?

Reply:
How to launch passive servers is not documented anywhere in the manual. The practical difficulty is that the launching thread must know when the to-be passive thread is done initializing and is blocking on IPC so it can safely unbind the initial SC.  If it unbinds too soon, the thread will be stuck forever. So some kind of startup handshake protocol is needed between the launcher task and the just launched passive task.

When binding a SC to both the passive task and a notification object the task will wait for, no such handshake is necessary, as the SC will be unbound from the TCB when the task blocks on the bound notification object. Of course this is more useful for passive servers handling notifications to begin with, but it can also simplify launching other passive servers (at the cost of a notification object).

The same for timeout exceptions: With this feature you can just bind a SC to the TCB when the task runs out of time, and the SC will be automatically "returned" when done (blocking on the notification again). Otherwise you have the whole startup problem again, where the passive task needs to know when to tell the monitor tasks that it's safe to unbind the SC.  But this time you can't implement it simply and synchronously.

My intention was to point the user in the right direction, without going into all the details.

Reply:
But why not put the explanation you wrote above here into the manual to have something at least. We can still move this around at a later point in time.

Reply:
I think the manual could use some appendices with code examples with details like this further explained.

Personally I don't like the tutorials because they're made as exercises and not very useful if you just want information. The solutions do give some additional information, but they are hard to get (at least I haven't found a link on the website for them, you need a local checkout AFAIK).

Should I remove the "This is useful when launching passive servers or handling timeout exceptions." altogether then?

Reply:
Personally, I'm Ok with it in, but it's hard for me to judge what people get confused about when they read it in context. It's more information to have it in. It's potentially less confusion if we remove it.

@axel-h what do you think? I can understand that it sparks the desire for more information, but this spot is not a good place to provide that information. Better not to have it at all then?

Fully agree that code examples in the manual would be good. Then this could be an easy reference to one of those. More work for the future..

Reply:
let's leave it as it is.